Fix invalid expect block and support for multi-line errors

GreasySlug opened the pull request:
Fixes #58  and #248.

Each line of the REPL is now evaluated by the `Parser::eval()` (Parser and Lex are not modified)

Changes proposed in this PR:
- Add Runnable Parser to check invalid syntax
- Check invalid block as a error
- Add block_begin to use for REPL
- Display multi-line error

@mtshiba

GreasySlug commented on compiler/erg_common/error.rs:
```diff
@@ -45,6 +45,7 @@ pub enum ErrorKind {
     VisibilityError = 20,
     MethodError = 21,
     DummyError = 22,
+    ExpectNewLine = 23,
```
I've tentatively added the error numbers.
Please comment if the right number and location is available.

mtshiba replied:
👍

GreasySlug commented:
Currently it is implemented only in `dummy.rs`.
Is it necessary to implement `run()` other than `dummy.rs`?

mtshiba commented:
A method should be added to the `Runnable` trait as a request method to determine if the next line should be followed. The name should be something like `expect_nextline()`.

GreasySlug commented on src/dummy.rs:
```diff
@@ -176,6 +180,108 @@ impl Runnable for DummyVM {
+                output.write_all(instance.ps1().as_bytes()).unwrap();
+                output.flush().unwrap();
+                let mut in_block = false;
+                let mut parser = ParserRunner::new(cfg);
```
parser is declared here, and I am using it, but it is not possible to add this to the Runnable trait.
Or is there a way around this?

mtshiba replied:
`expect_nextline` should be declared with no body, and the specific judgment algorithm should be defined in each struct. Compiler, Parser etc. uses Parser, Lexer should simply return `false`.

GreasySlug replied:
Thank you!!
I'll try it

GreasySlug commented:
I fixed conflict.

GreasySlug commented:
![multi-line-error](https://user-images.githubusercontent.com/106496265/204439703-e9d1cb09-8e79-44cc-9ee7-54130ce9a1f1.png)

GreasySlug commented on compiler/erg_common/traits.rs:
```diff
@@ -347,37 +347,38 @@ pub trait LimitedDisplay {
+pub enum BlockKind {
+    Main,          // now_block Vec must contain this
+    Assignment,    // =
+    Lambda,        // =>, ->, do:, do!:
```
I have included the `:colon` in the Lambda
Is there a better way to handle this?

mtshiba replied:
you mean like this?

```python
match! x:
    ...
```

Defining an additional enum option with a name like `ColonCall` may be better.

GreasySlug replied:
Thank you

GreasySlug commented on compiler/erg_parser/parse.rs:
```diff
@@ -905,14 +926,24 @@ impl Parser {
+            let body = self.try_reduce_block().map_err(|_| {
+                self.stack_dec();
+                if let Some(err) = self.errs.last_mut() {
+                    err.set_hint("Lambda");
```
When an error occurs the last element is available for extraction
Extract it and have a string as a hint

GreasySlug commented:
I fixed #281 then I found some bugs.
so I need more time to test and fix them

mtshiba commented:
When REPL detects an error, it refers to the previous source information.

<img width="548" alt="スクリーンショット 2024-02-28 20 07 50" src="https://github.com/erg-lang/erg/assets/45118249/07bcda60-03ca-41ef-a021-1e26bfa95686">

GreasySlug commented:
I addressed the issue occurring during single-line errors.

Additionally, I added error handling for cases where newline are expected, such as with `do!:`, `do:`, to ensure evaluation in the REPL.

![error](https://github.com/erg-lang/erg/assets/106496265/8d560859-711a-4f5f-a917-9f399d7282ae)

mtshiba commented:
REPL references previous source information when dealing with multiple lines of code.

<img width="459" alt="スクリーンショット 2024-02-29 13 30 07" src="https://github.com/erg-lang/erg/assets/45118249/4a393fb6-b20d-4da2-91ab-0d27db68ff23">

<img width="454" alt="スクリーンショット 2024-02-29 13 31 46" src="https://github.com/erg-lang/erg/assets/45118249/951ffb8d-2784-4de4-9d3a-d77c0b510259">

GreasySlug commented:
- Multi line errors
![error](https://github.com/erg-lang/erg/assets/106496265/6b433ceb-9d17-4773-8b1c-209d26da78a7)

- Collections
![collections](https://github.com/erg-lang/erg/assets/106496265/fc436d4f-8cb8-4144-96e9-cf4e9c1edf9d)

mtshiba commented:
Thank you!
